### ISSUE TITLE
Support for isFinite()

### ISSUE THREAD

**Issue by JetStarBlues** [Khan/live-editor#450]
[isFinite()](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/isFinite) does not seem to work in the Khan Academy pjs environment.
Error buddy says `__env__.isFinite is not a function`.

If try to use as `Number.isFinite()`, 
error buddy says `Cannot read property 'isFinite' of undefined`

Possible use case for isFinite() is a check for division by zero.

**Comment by GoToLoop** [comment]
I wonder whether **window.isFinite()** would cheat the "Oh noes"?

**Comment by JavascriptFTW** [comment]
This is a simple fix. Just mess with some stuff in `pjs-ast-transforms.js` to tell it to not prefix `isFinite`.

**Comment by JavascriptFTW** [comment]
Scratch that. Looking around it would be more consistent to add it in `pjs-code-injector.js`

**Comment by kevinbarabash** [comment]
Yup.  If someone wants to submit a PR for this it would be much appreciated.  The change should happen here https://github.com/Khan/live-editor/blob/master/js/output/pjs/pjs-ast-transforms.js#L118-L120.  @JavascriptFTW what change were you thinking of making in `pjs-code-injector.js`?

**Comment by JavascriptFTW** [comment]
@kevinbarabash I was saying that it seems to be more consistent with the style of the project to fix this bug by editing `pjs-code-injector.js` rather than `pjs-ast-transform.js` since the latter is dealing with constants and the former is actually attaching global methods to the environment (which is what this issue needs)
For reference:
https://github.com/Khan/live-editor/blob/master/js/output/pjs/pjs-code-injector.js#L132

**Comment by kevinbarabash** [comment]
@JavascriptFTW good call!  I forgot that `isNaN` is copied there.

**Comment by JetStarBlues** [comment]
Thanks for looking into this!

**Comment by JetStarBlues** [comment]
> If someone wants to submit a PR for this it would be much appreciated

I know very little about the KA codebase...
Is the solution as simple as @JavascriptFTW 's suggestion of adding,
`this.processing.isFinite = window.isFinite;` to [here](https://github.com/Khan/live-editor/blob/master/js/output/pjs/pjs-code-injector.js#L132)?
If that's the case and if people are busy I can give it a go.

**Comment by Potato42** [comment]
Why don't we allow `Number`? It would be easier to fix by just allowing `Number` which would then have the `isFinite` method.

**Comment by pamelafox** [comment]
I think it'd be fine to add Number to the array of objects we expose. I
generally like moving towards supporting all of JS when possible. I've
added Math and Array to that array before and it should be a simple change.
Please include a test case as well.

**Comment by JetStarBlues** [comment]
So, I was going to do a pull request. But I noticed my build on Windows added a lot of unnecessary changes. See, https://github.com/JetStarBlues/live-editor/commit/ac9f95349e6d2815201d4b776f87b89ffbb0c74c
At the very end of all that mess are the two things I did change as per the discussion above.

---
The demos/simple works for Number.isInfinite(). But I was unable to run the gulp tests.
Instead I got this error when I ran `node node_modules/gulp/bin/gulp.js test`

	undefined:1
	Timeout: tests did not start. This may be because of errors on the page.
	^
	
	SyntaxError: Unexpected token T
	    at Object.parse (native)
	    at Socket.<anonymous> (C:\Users\JK\Desktop\live-editor\testutil\gulp-mocha-runner.js:139:30)
	    at emitOne (events.js:77:13)
	    at Socket.emit (events.js:169:7)
	    at readableAddChunk (_stream_readable.js:146:16)
	    at Socket.Readable.push (_stream_readable.js:110:10)
	    at Pipe.onread (net.js:523:20)

**Comment by kevinbarabash** [comment]
@JetStarBlues gulp-eol is supposed to write files with "\n" for line endings instead of "\r\n".  Fore some reason it doesn't seem to be working.

As for the "Timeout: tests did not start." issue it looks like it's either something to do with phantomjs... maybe how/where we're trying to load pages from... maybe the fact that we're not dealing with paths in a cross platform way.

**Comment by JetStarBlues** [comment]
I went ahead and installed Ubuntu and tried again https://github.com/JetStarBlues/live-editor/commit/f04e45eb179c1cc183aec37d2bc4234b918728d8
The docs say to include the build files ... but there seems to be an awful lot of unrelated code when you do this?

**Comment by JavascriptFTW** [comment]
@JetStarBlues Wow. What editor are you using? Your editor may be trying to do something funky with the files.

**Comment by JetStarBlues** [comment]
@JavascriptFTW Sublime Text. 
Will give it one more go from scratch. To double check, procedure is,

    git clone
    git submodule update --init --recursive 
    npm install
    bower install
    gulp
    // make your edits
    git commit

Thanks for your time!

**Comment by kevinbarabash** [comment]
@JetStarBlues verify that there are no changes to the build files after running `gulp`.

**Comment by JetStarBlues** [comment]
@kevinbarabash from just the build (and before making any changes to the code), `git status` shows that these files were modified,  
- build/js/live-editor.audio.js
- build/js/live-editor.ui.js
- build/tmpl/tipbar.js

**Comment by kevinbarabash** [comment]
@JetStarBlues sorry, I missed this comment.  It shouldn't be updating those files.  I did a fresh checkout and ran the commands in your previous comment and those files didn't change for me.  Could you try again with a fresh clone?

**Comment by JetStarBlues** [comment]
@kevinbarabash Just did a fresh clone and build with the same commands. Still without modifying a single file, `git status` shows,

	modified:   build/js/live-editor.audio.js
	modified:   build/js/live-editor.debugger.js
	modified:   build/js/live-editor.output_sql.js
	modified:   build/js/live-editor.tooltips.js
	modified:   build/js/live-editor.ui.js
	modified:   build/tmpl/debugger.js
	modified:   build/tmpl/dev-record.js
	modified:   build/tmpl/image-picker.js
	modified:   build/tmpl/live-editor.js
	modified:   build/tmpl/mediapicker-modal.js
	modified:   build/tmpl/mediapicker-preview.js
	modified:   build/tmpl/sql-results.js
	modified:   build/tmpl/tipbar.js

I am running Ubuntu 15.04

**Comment by kevinbarabash** [comment]
I had a look at the repo you linked to.  It looks like a change in how the templates are being compiled.  Can you look in `node_modules` and let me know which version of `handlebars` was installed?  I also, tried deleting all of the build files and rebuilding locally.  This initially caused differences in the following files:
```
        modified:   build/js/live-editor.audio.js
	modified:   build/js/live-editor.debugger.js
	modified:   build/js/live-editor.output_sql.js
	modified:   build/js/live-editor.tooltips.js
	modified:   build/js/live-editor.ui.js
	modified:   build/tmpl/tipbar.js
```
This is because the templates are built after the `build/js/*.js` files which include them.  We should change the order.  Then I run `gulp build` a second time and the following files had differences:
```
	modified:   build/js/live-editor.ui.js
	modified:   build/tmpl/tipbar.js
```
And the only difference in the template was:
```
--- i/build/tmpl/tipbar.js
+++ w/build/tmpl/tipbar.js
@@ -7,7 +7,7 @@ this["Handlebars"]["templates"]["tipbar"] = Handlebars.template(function (Handle
 function program1(depth0,data) {
   
   
-  return "x";}
+  return "&times;";}
 
 function program3(depth0,data) {
```
which I think makes sense.

**Comment by JetStarBlues** [comment]
Version 1.0.5-beta was installed.
Not sure how to go about changing the build order...

**Comment by kevinbarabash** [comment]
I can look into that... it's still weird that we're getting different results from building the same thing.  I'm on Mac OS X, but I don't think that would cause any difference.  What version of npm are you using?

**Comment by JetStarBlues** [comment]
My npm version is 2.14.20

**Comment by kevinbarabash** [comment]
I'm really grasping at straws here.  I can't see a reason why this would affect things, but give it a try.

**Comment by JetStarBlues** [comment]
Here's what I got after updating npm (3.8.1) and bower (1.7.7), then doing a fresh clone and build.
https://github.com/JetStarBlues/Ignore_ReferenceLiveEditor3/commit/7347c0e8900465cbed1e3131794a82a42509e976
Also below are the gulp notifications if relevant.

Are other people having success with builds? If it's just my machine, someone else can commit the change.

---
	[19:54:32] Using gulpfile ~/Desktop/live-editor/gulpfile.js
	[19:54:32] Starting 'watch'...
	[19:54:33] Finished 'watch' after 479 ms
	[19:54:33] Starting 'check'...
	all files exist
	[19:54:33] Finished 'check' after 2.23 ms
	[19:54:33] Starting 'templates'...
	[19:54:33] Finished 'templates' after 37 ms
	[19:54:33] Starting 'script_core_deps'...
	[19:54:33] Starting 'script_editor_ace_deps'...
	[19:54:33] Starting 'script_editor_ace'...
	[19:54:33] Starting 'script_editor_textarea'...
	[19:54:33] Starting 'script_editor_structured_blocks_deps'...
	[19:54:33] Starting 'script_editor_structured_blocks'...
	[19:54:33] Starting 'script_audio'...
	[19:54:33] Starting 'script_tooltips'...
	[19:54:33] Starting 'script_debugger'...
	[19:54:33] Starting 'script_shared'...
	[19:54:33] Starting 'script_output_pjs_deps'...
	[19:54:33] Starting 'script_output_debugger_deps'...
	[19:54:33] Starting 'script_output_pjs'...
	[19:54:33] Starting 'script_output_webpage_deps'...
	[19:54:33] Starting 'script_output_sql_deps'...
	[19:54:33] Starting 'script_output_webpage'...
	[19:54:33] Starting 'script_output_sql'...
	[19:54:33] Starting 'script_output'...
	[19:54:33] Starting 'script_ui'...
	[19:54:33] Starting 'workers'...
	[19:54:33] Finished 'workers' after 5.79 ms
	[19:54:33] Starting 'style_core_deps'...
	[19:54:33] Starting 'style_editor_structured_blocks'...
	[19:54:33] Starting 'style_audio'...
	[19:54:33] Starting 'style_tooltips'...
	[19:54:33] Starting 'style_output'...
	[19:54:33] Starting 'style_ui'...
	[19:54:33] Starting 'style_debugger'...
	[19:54:33] Starting 'images'...
	[19:54:33] Finished 'images' after 2.11 ms
	[19:54:33] Starting 'externals'...
	[19:54:33] Finished 'externals' after 3.04 ms
	[19:54:36] Starting 'templates'...
	[19:54:36] Finished 'templates' after 2.86 ms
	[19:54:36] Finished 'script_editor_structured_blocks' after 3.37 s
	[19:54:36] Finished 'script_editor_textarea' after 3.39 s
	[19:54:36] Finished 'script_editor_ace' after 3.4 s
	[19:54:36] Finished 'style_core_deps' after 3.29 s
	[19:54:36] Finished 'style_audio' after 3.28 s
	[19:54:36] Finished 'style_output' after 3.28 s
	[19:54:36] Finished 'style_debugger' after 3.28 s
	[19:54:37] Starting 'templates'...
	[19:54:37] Finished 'templates' after 2.14 ms
	[19:54:37] Finished 'script_output_webpage' after 4.1 s
	[19:54:37] Finished 'style_ui' after 4.13 s
	[19:54:38] Finished 'script_output_sql_deps' after 4.44 s
	[19:54:38] Starting 'templates'...
	[19:54:38] Finished 'templates' after 2.17 ms
	[19:54:38] Finished 'script_debugger' after 4.79 s
	[19:54:38] Finished 'script_output' after 4.74 s
	[19:54:38] Finished 'script_output_sql' after 4.75 s
	[19:54:38] Finished 'style_editor_structured_blocks' after 4.72 s
	[19:54:38] Starting 'templates'...
	[19:54:38] Finished 'templates' after 2.68 ms
	[19:54:38] Finished 'style_tooltips' after 5.11 s
	[19:54:38] Starting 'styles'...
	[19:54:38] Finished 'styles' after 12 μs
	[19:54:38] Finished 'script_output_debugger_deps' after 5.17 s
	[19:54:39] Finished 'script_audio' after 6.16 s
	[19:54:40] Starting 'templates'...
	[19:54:40] Finished 'templates' after 2.94 ms
	[19:54:40] Finished 'script_output_pjs' after 6.55 s
	[19:54:40] Finished 'script_ui' after 6.51 s
	[19:54:40] Finished 'script_editor_structured_blocks_deps' after 6.7 s
	[19:54:40] Finished 'script_editor_ace_deps' after 6.72 s
	[19:54:40] Finished 'script_shared' after 6.76 s
	[19:54:40] Starting 'templates'...
	[19:54:40] Finished 'templates' after 1.64 ms
	[19:54:40] Finished 'script_core_deps' after 7.42 s
	[19:54:40] Finished 'script_output_pjs_deps' after 7.33 s
	[19:54:41] Finished 'script_output_webpage_deps' after 7.38 s
	[19:54:41] Starting 'templates'...
	[19:54:41] Finished 'templates' after 2.21 ms
	[19:54:41] Starting 'templates'...
	[19:54:41] Finished 'templates' after 4.29 ms
	[19:54:41] Starting 'templates'...
	[19:54:41] Finished 'templates' after 1.46 ms
	[19:54:41] Starting 'script_output_sql'...
	[19:54:41] Starting 'templates'...
	[19:54:41] Finished 'templates' after 1.61 ms
	[19:54:41] Starting 'script_ui'...
	[19:54:41] Finished 'script_tooltips' after 7.91 s
	[19:54:41] Finished 'script_output_sql' after 454 ms
	[19:54:42] Finished 'script_ui' after 906 ms
	[19:54:42] Starting 'scripts'...
	[19:54:42] Finished 'scripts' after 14 μs
	[19:54:42] Starting 'build'...
	[19:54:42] Finished 'build' after 11 μs
	[19:54:42] Starting 'default'...
	[19:54:42] Finished 'default' after 11 μs

**Comment by MatthiasSaihttam** [comment]
Can be closed as #585 has been merged.
